source/docs/advanced/graphql/slim-down-resolvers-with-loaders.md

```diff
@@ -0,0 +1,351 @@
+---
+id: slim-down-resolvers-with-loaders
+title: Slim down resolvers with loaders
+---
+
+GraphQL modules may grow over time depending on the features they provide, and
+remote services they interact with. GraphQL resolvers will then have more and
+more code, and very often require developers to introduce an abstraction
+**containing all the business logic to keep code maintainable and increase
+testability.**
+
+In Front-Commerce’s core and platforms integrations (such as Magento2) we have
+introduced the concept of _« loaders »_ to achieve this.
+
+This page explains what loaders are, how you can create them and should allow
+you to have a better understanding of our code when browsing it.
+
+<blockquote class="info">
+  If you have used Apollo Server recently, you might already be familiar with a
+  similar pattern named [data
+  sources](https://www.apollographql.com/docs/tutorial/data-source.html). It is
+  also similar to [GrAMPS’s
+  models](https://gramps-graphql.github.io/gramps-express/data-source/tutorial-model/).
+</blockquote>
+
+## Why loaders?
+
+When resolvers grow, you will likely extract functions or modules to share code
+between resolvers. It could be constructing an `axios` instance with the correct
+url and headers, modifying responses to return objects matching the GraphQL
+schema or error management for instance.
+
+Loaders are aimed at being node modules encapsulating all your business specific
+code. We recommend to design them as pure JS modules without any dependency on
+Front-Commerce or GraphQL features. Designing loaders with this approach has
+several advantages:
+
+- you can reuse existing code or libraries from another project
+- if you migrate away from Front-Commerce, this code could be reused easily in
+  your new application (as soon as it is in a Javascript environment of course)
+- testability: the code from both your resolvers and loaders would be easier to
+  test. You could inject fakes or mock loaders to test resolvers, and you could
+  test loaders without the complexity of a GraphQL context since they are pure
+  JS modules.
+- code sharing: under some circumstances you may also find convenient to share
```

Comment:
Is that for loaders or other things like adapters?  
Because here I understand it's about loaders, but it would mean we need to create a dataloader on the client side, etc.

I'm not quite sure this is the best idea.

Reply:
It was more about the fact that loaders were pure JS code: it could lead to refactoring parts in smaller modules that could be shared.

This is not the best point of all of them. We will see if it leads to problematic implementations or poor practices.